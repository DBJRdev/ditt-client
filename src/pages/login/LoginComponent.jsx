import PropTypes from 'prop-types';
import React from 'react';
import decode from 'jwt-decode';
import { Link } from 'react-router-dom';
import { withTranslation } from 'react-i18next';
import {
  Button,
  Card,
  CardBody,
  Alert,
  TextField,
} from '@react-ui-org/react-ui';
import {
  Icon,
  LoadingIcon,
} from '../../components/Icon';
import {
  ROLE_ADMIN,
  ROLE_EMPLOYEE,
  ROLE_SUPER_ADMIN,
} from '../../resources/user';
import routes from '../../routes';
import styles from './Login.scss';
import LogoImage from './images/logo.svg';

class LoginComponent extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      password: null,
      username: null,
    };

    this.onChangeHandler = this.onChangeHandler.bind(this);
    this.onSubmitHandler = this.onSubmitHandler.bind(this);
  }

  componentDidMount() {
    if (this.props.token) {
      const decodedToken = decode(this.props.token);

      if (decodedToken) {
        const { roles } = decodedToken;

        if (roles.includes(ROLE_ADMIN) || roles.includes(ROLE_SUPER_ADMIN)) {
          this.props.history.push(routes.userList);
        } else if (roles.includes(ROLE_EMPLOYEE)) {
          this.props.history.push(routes.index);
        }
      }
    }
  }

  onChangeHandler(e) {
    this.setState({
      [e.target.id]: e.target.value,
    });
  }

  onSubmitHandler(e) {
    const {
      login,
      resetLogoutLocally,
    } = this.props;
    const {
      password,
      username,
    } = this.state;

    e.preventDefault();

    resetLogoutLocally();
    login({
      password,
      username,
    });
  }

  render() {
    const {
      isPosting,
      isPostingFailure,
      isLoggedOutLocally,
      t,
    } = this.props;
    const {
      password,
      username,
    } = this.state;

    return (
      <div className={styles.container}>
        <LogoImage
          alt={t('layout:title')}
          className={styles.logo}
          height={141}
          width={302}
        />
        <div className={styles.box}>
          <div className={styles.title}>
            {t('layout:title')}
          </div>
          {(isPostingFailure || isLoggedOutLocally) && (
            <div className="mb-5">
              <Alert
                icon={<Icon icon="error" />}
                type="error"
              >
                <strong>
                  {t('general:text.error')}
                  {': '}
                </strong>
                {
                  isLoggedOutLocally
                    ? t('login:text.loggedOutAutomatically')
                    : t('login:validation.invalidUsernameOrPassword')
                }
              </Alert>
            </div>
          )}
          <Card variant="bordered">
            <CardBody>
              <form
                onSubmit={this.onSubmitHandler}
              >
                <div className="mb-3">
                  <TextField
                    autoComplete="username"
                    changeHandler={this.onChangeHandler}
                    fullWidth
                    id="username"
                    label={t('user:element.email')}
                    type="email"
                    required
                    value={username ?? ''}
                  />
                  <TextField
                    autoComplete="current-password"
                    changeHandler={this.onChangeHandler}
                    fullWidth
                    id="password"
                    label={t('user:element.plainPassword')}
                    type="password"
                    required
                    value={password ?? ''}
                  />
                </div>
                <Button
                  block
                  id="signInButton"
                  label={t('login:action.signIn')}
                  loadingIcon={isPosting && <LoadingIcon />}
                  type="submit"
                />
              </form>
              <div className={styles.footer}>
                <Link to={routes.forgotPassword}>
                  {t('login:action.forgotPassword')}
                </Link>
              </div>
            </CardBody>
          </Card>
        </div>
      </div>
    );
  }
}

LoginComponent.defaultProps = {
  token: null,
};

LoginComponent.propTypes = {
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }).isRequired,
  isLoggedOutLocally: PropTypes.bool.isRequired,
  isPosting: PropTypes.bool.isRequired,
  isPostingFailure: PropTypes.bool.isRequired,
  login: PropTypes.func.isRequired,
  resetLogoutLocally: PropTypes.func.isRequired,
  t: PropTypes.func.isRequired,
  token: PropTypes.string,
};

export default withTranslation()(LoginComponent);
