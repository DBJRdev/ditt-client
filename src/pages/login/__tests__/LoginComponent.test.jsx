import React from 'react';
import { mount } from 'enzyme';
import { StaticRouter } from 'react-router-dom';
import LoginComponent from '../LoginComponent';

describe('rendering', () => {
  it('renders correctly', () => {
    const props = {
      history: {
        push: () => {},
      },
      isLoggedOutLocally: false,
      isPosting: false,
      isPostingFailure: false,
      login: () => {},
      resetLogoutLocally: () => {},
    };

    const tree = mount((
      <StaticRouter context={{}} location="/path">
        <LoginComponent {...props} />
      </StaticRouter>
    ));

    expect(tree).toMatchSnapshot();
  });
});
