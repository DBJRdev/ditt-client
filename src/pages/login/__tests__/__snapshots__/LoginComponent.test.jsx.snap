// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly 1`] = `
<div
  className="container"
>
  <svgr-mock
    alt="layout:title"
    className="logo"
    height={141}
    width={302}
  />
  <div
    className="box"
  >
    <div
      className="title"
    >
      layout:title
    </div>
    <div
      className="Card__root__2TI9U7gf
        
        Card__isVariantBordered__1CPRD7Sx"
    >
      <div
        className="Card__body__VVBN5EiS"
      >
        <form
          onSubmit={[Function]}
        >
          <div
            className="mb-3"
          >
            <label
              className="TextField__root__3iUhzvuP TextField__isRootFullWidth__WsG8m8iX   TextField__rootLayoutVertical__1gRnSdx1  TextField__isRootRequired__35S_G0GS TextField__rootSizeMedium__2bbQl9zu  TextField__rootVariantOutline__3NOXZDWS"
              htmlFor="username"
              id="username__label"
              style={Object {}}
            >
              <div
                className="TextField__label__38MaVqvq "
                id="username__labelText"
              >
                user:element.email
              </div>
              <div
                className="TextField__field__2IQ3hvDk"
              >
                <div
                  className="TextField__inputContainer__zi07SHRD"
                >
                  <input
                    autoComplete="username"
                    className="TextField__input__2lIhECsc"
                    disabled={false}
                    id="username"
                    max={null}
                    onChange={[Function]}
                    placeholder={null}
                    required={true}
                    size={null}
                    type="email"
                    value=""
                  />
                </div>
              </div>
            </label>
            <label
              className="TextField__root__3iUhzvuP TextField__isRootFullWidth__WsG8m8iX   TextField__rootLayoutVertical__1gRnSdx1  TextField__isRootRequired__35S_G0GS TextField__rootSizeMedium__2bbQl9zu  TextField__rootVariantOutline__3NOXZDWS"
              htmlFor="password"
              id="password__label"
              style={Object {}}
            >
              <div
                className="TextField__label__38MaVqvq "
                id="password__labelText"
              >
                user:element.plainPassword
              </div>
              <div
                className="TextField__field__2IQ3hvDk"
              >
                <div
                  className="TextField__inputContainer__zi07SHRD"
                >
                  <input
                    autoComplete="current-password"
                    className="TextField__input__2lIhECsc"
                    disabled={false}
                    id="password"
                    max={null}
                    onChange={[Function]}
                    placeholder={null}
                    required={true}
                    size={null}
                    type="password"
                    value=""
                  />
                </div>
              </div>
            </label>
          </div>
          <button
            className="Button__root__1cQ7pn3O
        Button__priorityFilled__254OCUyY
        Button__sizeMedium__3SsHOYL-
        Button__variantPrimary__2utVPMdx
        Button__isRootBlock__3-DBqjBK"
            disabled={false}
            id="signInButton"
            onClick={null}
            type="submit"
          >
            <span
              className="Button__label__3pJOzZqB"
              id="signInButton__labelText"
            >
              login:action.signIn
            </span>
          </button>
        </form>
        <div
          className="footer"
        >
          <a
            href="/forgot-password"
            onClick={[Function]}
          >
            login:action.forgotPassword
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
